### ISSUE TITLE
HTTPPageGetter handleStatus_301 - Improper Redirect

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3624]
|<img alt="darkops's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| darkops reported|
|-|-|
|Trac ID|trac#3624|
|Type|enhancement|
|Created|2009-01-18 19:13:21Z|
|Branch|https://github.com/twisted/twisted/tree/deviate-redirects-3624|

If a user makes a request using the POST method and the server responds with a any 3xx redirect, any class using the HTTPPageGetter class will follow the redirect with a subsequent POST request. 

If the location field in the server's response includes GET parameters the server only accepts through GET requests, the server will see the clients POST request and ignore all parameters in the path. 

This can be easily fixed by simply inserting the following line into HTTPPageGetter.handleStatus_301 line ![#7441](https://github.com/twisted/twisted/issues/7441):

self.factory.method = 'GET'

To further illustrate the problem and my fix, included below is a simple demonstration:

**Original Request**
```
POST /login_page.adp HTTP/1.1
HOST: SampleDomain.com
Content-Length: 78


login=true&email=elite@...&password=123456&submitbutton=Login
```

**Server Responds with 3xx Redirect**
```
HTTP/1.0 302 Found
Location: new_page.adp?email=elite%40mytrainingground%2ecom&h=4b90f8207a0dc2c4b12a45c096fae25&action=getacceptance&destination=final_page.adp
MIME-Version: 1.0
Content-Type: text/html
Content-Length: 408
Connection: close
```

**handleStatus_301 Creates the redirect request using the POST method**
```
POST new_page.adp?email=elite%40mytrainingground%2ecom&h=4b90f8207a0dc2c4b12a45c096fae25&action=getacceptance&destination=final_page.adp HTTP/1.1
HOST: SampleDomain.com
Content-Length: 0
```

**After inserting the recommended line into handleStatus_301, the redirect request is correctly made using the GET method**
```
GET new_page.adp?email=elite%40mytrainingground%2ecom&h=4b90f8207a0dc2c4b12a45c096fae25&action=getacceptance&destination=final_page.adp HTTP/1.1
HOST: SampleDomain.com
```

Attachments:

* [nonRFC2616CompliantRedirect.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2c7/2c72e172a2fb4460b0a8f6640e8565a903eb8bd7/f5a43c6f01894398d40601a46a77a3a0eb0b3b0c.patch) (5336 bytes) - added by truekonrads on 2009-08-12 17:54:45Z - Updated version of original patch
* [nonRFC2616CompliantRedirect.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/2c7/2c72e172a2fb4460b0a8f6640e8565a903eb8bd7/86cd27dff93016aa32b429c0ec7d167bd544668f.patch) (5282 bytes) - added by truekonrads on 2009-08-12 18:12:11Z - Some minor improvements.

<details><summary>Searchable metadata</summary>

```
trac-id__3624 3624
type__enhancement enhancement
reporter__darkops darkops
priority__normal normal
milestone__ 
branch__branches_deviate_redirects_3624 branches/deviate-redirects-3624
branch_author__jonathanj jonathanj
status__closed closed
resolution__fixed fixed
component__web web
keywords__httpclient httpclient
time__1232306001000000 1232306001000000
changetime__1251415446000000 1251415446000000
version__None None
owner__jonathanj jonathanj
cc__chad@... cc__ralphm cc__konrads@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun set owner to @exarkun|
|-|-|

[The HTTP 1.1 RFC](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.3) clearly defines the correct behavior for this case.  However, many real user agents deviate from the RFC in this area, switching to `GET` erroneously for a `301` response.  `twisted.web.client` is doing what is correct according to the RFC, but not what agrees with other popular HTTP clients.  Adding a way to _enable_ this behavior (but leaving the current behavior by default) may be desirable.

Note also that the `Location` values in the example above are invalid.  The value of the `Location` header **must** be an absolute URI (<http://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.30>).

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_2"></a>|@ralphm commented|
|-|-|

Actually, per that spec, no client should automatically redirect upon receiving a `301` as a response to a request method other than `GET` or `HEAD`. For a `303`, however, changing the method to `GET` and doing the redirect automatically is proper.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

[#3952](https://github.com/twisted/twisted/issues/3952) was a duplicate of this.

**Comment by twisted-trac** [comment]
|<img alt="truekonrads's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|truekonrads commented|
|-|-|

Hi,

Is there a consensus or this 7 month old bug is a WONTFIX?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

No, there isn't.  There was discussion about the appropriate action to take.  Now the ticket is just waiting for someone to take it.

**Comment by twisted-trac** [comment]
|<img alt="truekonrads's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|truekonrads commented|
|-|-|

I have attached a patch that acts on 302 Found as a real web browser would. This is also in RFC 2616 spec discussed under [302 Found section](http://www.w3.org/Protocols/rfc2616/rfc2616-sec10.html#sec10.3.3).

**Comment by twisted-trac** [comment]
|[<img alt="ralphm's avatar" src="https://avatars.githubusercontent.com/u/116741?s=50" width="50" height="50">](https://github.com/ralphm)<a name="note_7"></a>|@ralphm set owner to truekonrads|
|-|-|

Thanks for the patch! Although this ticket was not put in review, I do have some comments and nits:

 * Pretty straightforward solution. Great!
 * As rfc:2616 mentions that many browsers handle a 302 status with a non-GET method as if a 303 was received, why not actually call the `handleStatus_303` method in that case?
 * `ExtendedRedirect` seems quite eloborate for what you want to test. I expected a very small resource that just always returns a 302 status.
 * `ExtendedRedirect` lacks docstrings and does not follow our coding standards w.r.t. spacing.
 * There should also be a test for the case where `nonRFC2616CompliantRedirect` is False.
 * There should be a test for the default case, either by testing the instance variable or the desired behaviour.
 * Line 540 of `test_webclient.py` misses a space after the comma.
 * I would not refer to this ticket in the comments, but just explain the reasoning in the docstring of the new instance variable and point to the RFC. You can then refer to this from other places with a `@see`.

Finally I'd like to point to the [extended description of the 302 status code in the current draft of httpbis](http://tools.ietf.org/html/draft-ietf-httpbis-p2-semantics-07#section-8.3.3), which introduces the concept of _safeness_.

Also, with this draft being actively worked on, maybe the new flag should not include the RFC number but be more descriptive.

**Comment by twisted-trac** [comment]
|<img alt="truekonrads's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|truekonrads set status to `assigned`|
|-|-|

Replying to [ralphm](#note_7):
> Thanks for the patch! Although this ticket was not put in review, I do have some comments and nits:
>
>  * Pretty straightforward solution. Great!
**blush**
>  * As rfc:2616 mentions that many browsers handle a 302 status with a non-GET method as if a 303 was received, why not actually call the `handleStatus_303` method in that case?
Done.
>  * `ExtendedRedirect` seems quite eloborate for what you want to test. I expected a very small resource that just always returns a 302 status.
I reduced the linecount a little. I don't think spending more time reducing a mock object is reasonable. If You can show an simpler way, please do - I learn from Twisted a lot. 
>  * `ExtendedRedirect` lacks docstrings and does not follow our coding standards w.r.t. spacing.
Cleaned up spacing, added docstring.
>  * There should also be a test for the case where `nonRFC2616CompliantRedirect` is False.
Done.
>  * There should be a test for the default case, either by testing the instance variable or the desired behaviour.
Done.
>  * Line 540 of `test_webclient.py` misses a space after the comma.
Auto-formatted the code.
>  * I would not refer to this ticket in the comments, but just explain the reasoning in the docstring of the new instance variable and point to the RFC. You can then refer to this from other places with a `@see`.
Updated docstring.
>
> Finally I'd like to point to the [extended description of the 302 status code in the current draft of httpbis](http://tools.ietf.org/html/draft-ietf-httpbis-p2-semantics-07#section-8.3.3), which introduces the concept of _safeness_.
>
> Also, with this draft being actively worked on, maybe the new flag should not include the RFC number but be more descriptive.
I renabled the parameter to enableUnsafeRedirects. I believe it is more in line with this concept. Thanks for the tip!

**Comment by twisted-trac** [comment]
|<img alt="truekonrads's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|truekonrads set owner to @ralphm<br>truekonrads set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_11"></a>|@jonathanj commented|
|-|-|

(In [27227]) Branching to 'deviate-redirects-3624'

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_12"></a>|@jonathanj set owner to truekonrads|
|-|-|

I've applied "nonRFC2616CompliantRedirect.2.patch" to a branch and cleaned up some style and documentation issues.

I also renamed `enabledUnsafeRedirects` (to `unsafeRedirects`) to better match the other parameters, however I'm wondering why the word "unsafe" was chosen. They're more of deviate redirects, aren't they?

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_13"></a>|@jonathanj commented|
|-|-|

(In [27230]) Match new parameter names in the tests, refs [#3624](https://github.com/twisted/twisted/issues/3624).

**Comment by twisted-trac** [comment]
|<img alt="truekonrads's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|truekonrads commented|
|-|-|

Replying to [jonathanj](#note_12):
> I've applied "nonRFC2616CompliantRedirect.2.patch" to a branch and cleaned up some style and documentation issues.
>
> I also renamed `enabledUnsafeRedirects` (to `unsafeRedirects`) to better match the other parameters, however I'm wondering why the word "unsafe" was chosen. They're more of deviate redirects, aren't they?
They are called unsafe, because according to httpbis, this action is [unsafe](http://tools.ietf.org/html/draft-ietf-httpbis-p2-semantics-07#section-8.3.3). According to standard, if, after a POST you get a 302 redirect, User Agent must not automatically follow, but ask for user interaction.

I think that leaving "unsafe redirects" is appropriate. The cleanup looks fine.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_15"></a>|@jonathanj removed owner|
|-|-|

I'm putting this up for review, so that someone else can take a look.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun set owner to @jonathanj<br>@exarkun set status to `new`|
|-|-|

This looks good to me.  I agree that the word "unsafe" isn't entirely appropriate here.  I had a conversation with truekonrads on IRC and I think I convinced him.  Unfortunately neither of us came up with a really good substitute. :)

A variety of suggestions were presented on IRC (and earlier on this ticket).  In chronological order:

  * deviateRedirects
  * safeRedirects
  * convenientButUnsafeRedirects
  * emulateBrowserFoundRedirectMethodRewrite
  * emulateBrowserRedirects
  * emulateBrowserRedirectBehavior
  * foundRedirectRequestUsesGET
  * standardConformantRedirects
  * getAfterFoundRedirect
  * afterFoundRedirectDoGet
  * getAfter301
  * after301Get
  * doGETAfter301
  * afterFoundGet

Note that all of the suggestions with 301 in their name are wrong and should use 302 instead.  I think the last one is the best of the lot.

Please just rename the flag and then (assuming you don't break anything ;) merge.

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_18"></a>|@jonathanj commented|
|-|-|

(In [27249]) Rename "unsafeRedirects" flag.

refs [#3624](https://github.com/twisted/twisted/issues/3624)

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_19"></a>|@jonathanj set status to `closed`|
|-|-|

(In [27250]) Merge "deviate-redirects-3624" branch.

Author: truekonrads
Reviewer: exarkun, jonathanj
Fixes: [#3624](https://github.com/twisted/twisted/issues/3624)

Introduce a parameter to HTTPClientFactory that enables redirection behaviour
commonly found in web browsers: following a POST with a 302 response status.
